### PR TITLE
[SMALLFIX] Fix ft failure window

### PR DESCRIPTION
**Pull request by aaudiber**
This PR fixes an issue where acked metadata changes could be lost if they happen immediately after the journal is closed. This issue was possible but extremely unlikely in previous versions. It only became a real problem after https://github.com/Alluxio/alluxio/pull/6628 merged two weeks ago.

1. RPC begins
2. Journal is closed, usually by the shutdown hook when stopping the Alluxio master, but possibly because the master loses connection to Zookeeper and steps down as primary.
3. RPC creates a journal context. The context is a NoopContext because the journal is closed.
4. RPC completes without writing a journal entry.
5. RPC is successful, but when master restarts the change is forgotten

To fix this, we should throw UnavailableException instead of returning a NoopContext when the journal is closed.

The interesting changes are in the "Fix bug where closing journal allows RPCs to skip journal write" and "Add test" commits.

### CODE REVIEW

**Comment by aaudiber** [comment]
Jenkins, test this please

File: core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournal.java [hunk 1]
```diff
@@ -146,36 +153,41 @@ public URI getLocation() {
...
   }
 
   @Override
-  public JournalContext createJournalContext() {
-    if (mAsyncWriter == null) {
-      return new NoopJournalContext();
+  public synchronized JournalContext createJournalContext() throws UnavailableException {
+    if (mState == State.CLOSED) {
+      // We throw UnavailableException here so that clients will retry with the next primary master.
+      throw new UnavailableException("Failed to write to journal: journal is closed");
```
Comment (calvinjia) [inline core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournal.java]: Should we throw this if the state is secondary?

File: core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournal.java [hunk 2]
```diff
@@ -146,36 +153,41 @@ public URI getLocation() {
...
+        "Must be in primary state to create journal contexts, but in state " + mState);
     return new MasterJournalContext(mAsyncWriter);
   }
 
-  private UfsJournalLogWriter writer() throws IOException {
-    if (mWriter == null) {
-      throw new IllegalStateException("Cannot write to the journal in secondary mode");
+  private synchronized UfsJournalLogWriter writer() throws IOException {
+    if (mState != State.PRIMARY) {
+      throw new IllegalStateException("Cannot write to the journal in state " + mState);
```
Comment (calvinjia) [inline core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournal.java]: Use a precondition here?

File: core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournal.java (hunk 1 above)
Comment (aaudiber) [reply core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournal.java]: done

File: core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournal.java (hunk 2 above)
Comment (aaudiber) [reply core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournal.java]: done

**Comment by aaudiber** [comment]
@calvinjia The class is threadsafe now - all methods which mutate state or access mutable state are `synchronized`

**Pull request merged**